### PR TITLE
[NVIDIA] Add overwrite_with_gradient collection

### PR DESCRIPTION
**Pull request by kaixih**
This PR provides the global constant string of overwrite_with_gradient and a helper function to check if the variable belongs to the category.

There are four related PRs, and should be reviewed in this order:
current-->(1) https://github.com/google/praxis/pull/29
(2) https://github.com/google/paxml/pull/48
(3) https://github.com/google/praxis/pull/28
(4) https://github.com/google/paxml/pull/49

cc. @pjannaty @reedwm @nluehr @lukaszlew

### CODE REVIEW

File: praxis/base_layer.py [hunk 1]
```diff
@@ -190,6 +191,12 @@ def var_skip_lp_regularization(var_params: ParamsT) -> bool:
   )
 
 
+def var_overwrite_with_gradient(var_hparams: ParamsT) -> bool:
+  """Returns True if var_hparams overwrites with gradient."""
```
Comment (lukaszlew) [inline praxis/base_layer.py:195]: Please add a more detailed documentation, something like this:
```
Returns True if var_hparams is OWG (overwrite with gradient) variable.

OWG variables can be thought as semi-trainable. 
Trainable variables are updated by the learner `x = x + update(grad)`
Non-trainable variables are left alone `x = x`
OWG variables are `x=grad`.
I.e. after each training step, they will be overwritten by the the value of the gradient with respect to themselves. 
```

**Review by lukaszlew — APPROVED** [review]

**Comment by zhangqiaorjc** [comment]
Overall good design. One should read https://github.com/google/flax/blob/main/flax/linen/fp8_ops.py too for where the new_scale and amax_history are produced and passed raw gradients to be updated in gradient step (that's where the variable assignment logic is)

**Comment by zhangqiaorjc** [comment]
Just to spell this out more explicitly for posterity


```
@partial(custom_vjp, nondiff_argnums=(0,))
def in_qdq(compute_dtype, inp, scale, amax_history):
  qin, _, _ = qdq_and_return(
      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
  return qin

def in_qdq_fwd(compute_dtype, inp, scale, amax_history):
  qin, new_scale, new_amax_history = qdq_and_return(
      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
  return qin, (new_scale, new_amax_history)
```

The returned value of in `_qdq_bwd` `new_scale, new_amax_history` are the raw gradients of arguments `scale, amax_history` to `in_qdq`, which will show up in the optimizer as raw gradients that are then treated as pass-through via `var_overwrite_with_gradient=True`, and then `x=grad` is basically the `self.scale=new_scale` that one might expect in PyTorch style code in layers directly.

**Review by zhangqiaorjc — APPROVED** [review]
